plugins/BEdita/Core/src/Shell/StreamsShell.php

```diff
@@ -0,0 +1,82 @@
+<?php
+/**
+ * BEdita, API-first content management framework
+ * Copyright 2019 ChannelWeb Srl, Chialab Srl
+ *
+ * This file is part of BEdita: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * See LICENSE.LGPL or <http://gnu.org/licenses/lgpl-3.0.html> for more details.
+ */
+namespace BEdita\Core\Shell;
+
+use BEdita\Core\Model\Entity\Stream;
+use Cake\Console\Shell;
+use Cake\I18n\Date;
+use Cake\ORM\TableRegistry;
+
+/**
+ * Stream shell commands: removeOrphans
+ *
+ * @since 4.0.0
+ */
+class StreamsShell extends Shell
+{
+
+    /**
+     * {@inheritDoc}
+     *
+     * @codeCoverageIgnore
+     */
+    public function getOptionParser()
+    {
+        $parser = parent::getOptionParser();
+        $parser->addSubcommand('removeOrphans', [
+            'help' => 'remove obsolete/orphans streams and related files',
+            'parser' => [
+                'description' => [
+                    'Remove orphans streams.'
+                ],
+                'options' => [
+                    'days' => [
+                        'help' => 'Days to consider for stream research for orphans',
+                        'required' => false,
+                        'default' => 1,
+                    ],
+                ],
+            ]
+        ]);
+
+        return $parser;
+    }
+
+    /**
+     * Create a new Stream
+     *
+     * @return void
+     */
+    public function removeOrphans()
+    {
+        $days = (int)$this->param('days');
+        $streamsTable = TableRegistry::get('Streams');
+        $query = $streamsTable
+            ->find()
+            ->where([
+                'object_id IS NULL',
+                'created >=' => new \DateTime(sprintf('-%d days', $days)),
+            ]);
+        $n = $query->count();
+        if ($n > 0) {
```

Comment:
IMHO this makes no sense. There's absolutely no point in performing an extra query to decide if another query should be performed, given that the second query would affect no rows.

Also, I'm pretty sure that performing a delete query that way (`$this->Streams->query()->delete()` or `$this->Streams->deleteAll()`) would **NOT** invoke `beforeDelete` and `afterDelete` callbacks, and therefore the filesystem entries would not be deleted. This would clean up the database, but would leave the filesystem dirty, and the cleanup would become even harder.

My two cents:

```php
$query = $this->Streams->find()
    ->where([
        // ...
    ]);

$count = 0;
foreach ($query as $stream) { // No need to `if`: if the result set is empty, this part is not executed.
    $this->verbose(sprintf('Deleting stream %s...', $stream->id));
    $this->Streams->deleteOrFail($stream);
    $count++;
}
$this->out(sprintf('%d stream(s) deleted', $count));
```